sycl/doc/extensions/proposed/sycl_ext_intel_device_aspects.asciidoc

```diff
@@ -0,0 +1,313 @@
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+[comment]
+--
+_Add the following paragraph when this specification becomes "experimental"._
+
+There are important limitations with the DPC++ implementation of this
+experimental extension.  See the section below titled "Limitations with the
+experimental version" for a description of these limitations.
+--
+
+
+== Overview
+
+Most device aspects provide a way for an application to test whether a device
+supports a specific feature.  However, sometimes an application needs to
+conditionally execute code based on the specific model of the device.  This
+extension adds new aspects, for supported Intel devices, which can be used for
+this purpose.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_DEVICE_ASPECTS` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== New aspects
+
+This extension adds the following aspects to `sycl::aspect`:
+
+```
+namespace sycl {
+
+enum class aspect : /* unspecified */ {
+  ext_intel_x86_64,
+  ext_intel_gpu_bdw,
+  ext_intel_gpu_skl,
+  ext_intel_gpu_kbl,
+  ext_intel_gpu_cfl,
+  ext_intel_gpu_apl,
+  ext_intel_gpu_glk,
+  ext_intel_gpu_whl,
+  ext_intel_gpu_aml,
+  ext_intel_gpu_cml,
+  ext_intel_gpu_icllp,
+  ext_intel_gpu_ehl,
+  ext_intel_gpu_tgllp,
+  ext_intel_gpu_rkl,
+  ext_intel_gpu_adl_s,
+  ext_intel_gpu_rpl_s,
+  ext_intel_gpu_adl_p,
+  ext_intel_gpu_adl_n,
+  ext_intel_gpu_dg1,
+  ext_intel_gpu_acm_g10,
+  ext_intel_gpu_acm_g11,
+  ext_intel_gpu_acm_g12,
+  ext_intel_gpu_pvc,
+
+  ext_intel_gpu_8_0_0 = ext_intel_gpu_bdw,
+  ext_intel_gpu_9_0_9 = ext_intel_gpu_skl,
+  ext_intel_gpu_9_1_9 = ext_intel_gpu_kbl
+  ext_intel_gpu_9_2_9 = ext_intel_gpu_cfl,
+  ext_intel_gpu_9_3_0 = ext_intel_gpu_apl,
+  ext_intel_gpu_9_4_0 = ext_intel_gpu_glk,
+  ext_intel_gpu_9_5_0 = ext_intel_gpu_whl,
+  ext_intel_gpu_9_6_0 = ext_intel_gpu_aml,
+  ext_intel_gpu_9_7_0 = ext_intel_gpu_cml,
+  ext_intel_gpu_11_0_0 = ext_intel_gpu_icllp,
+  ext_intel_gpu_11_2_0 = ext_intel_gpu_ehl,
+  ext_intel_gpu_12_0_0 = ext_intel_gpu_tgllp,
+  ext_intel_gpu_12_10_0 = ext_intel_gpu_dg1
+};
+
+} // namespace sycl
+```
+
+The following table tells which version of this extension first included each
+of these aspects, and it provides a brief description of their meanings.
+
+[%header,cols="5,1,5"]
+|===
+|Aspect name
+|Added in version
+|Description
+
+|`ext_intel_x86_64`
```

Comment:
This table partly duplicates https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#aot-target-identification

Could we refactor supported targets into a separate document maybe and refer to it from other docs? That would make designs more maintainable.

Reply:
We need a list of the extended aspects in the extension specification, and I think it needs to be self-contained.  For example, the list needs to clearly state which version of the extension adds each aspect.

If we want to reduce duplication, perhaps we could change the table in [OptionalDeviceFeatures ](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#aot-target-identification)to point to this extension specification.  However, note that some entries in the OptionalDeviceFeatures table do not appear in this list of aspects and the spellings do not exactly match.

Reply:
> If we want to reduce duplication, perhaps we could change the table in [OptionalDeviceFeatures ](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#aot-target-identification)to point to this extension specification.

I agree.

> However, note that some entries in the OptionalDeviceFeatures table do not appear in this list of aspects and the spellings do not exactly match.

Maybe some wording could be added to OptionalDeviceFeatures - 1) which aspects have corresponding representation in the config file and `-fsycl-targets` 2) rule how this corresponding representation is obtained from aspect's string.

Reply:
Done in 02669efbf8e151d491a9bfe71d0301c3a129fb7b